contrib/pkg/awsprivatelink/endpointvpc/add.go

```diff
@@ -0,0 +1,469 @@
+package endpointvpc
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	hiveutils "github.com/openshift/hive/contrib/pkg/utils"
+	awsutils "github.com/openshift/hive/contrib/pkg/utils/aws"
+	"github.com/openshift/hive/pkg/awsclient"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type endpointVPCAddOptions struct {
+	associatedVpcs    []hivev1.AWSAssociatedVPC
+	endpointVpcRegion string
+	endpointVpcId     string
+	endpointSubnetIds []string
+
+	dynamicClient      client.Client
+	endpointVpcClients awsclient.Client
+	awsClientsByRegion map[string]awsclient.Client
+}
+
+func NewEndpointVPCAddCommand() *cobra.Command {
+	opt := &endpointVPCAddOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "add vpc-id",
+		Short: "Networking setup between an endpoint VPC and each associated VPC",
+		Long: `Networking setup between an endpoint VPC and each associated VPC 
+specified in HiveConfig.spec.awsPrivateLink.associatedVPCs: 
+1) Establish VPC peering connection between the endpoint VPC and the associated VPC
+2) Add a route (to the peered VPC) in relevant route tables of the associated VPC and the endpoint VPC 
+3) Update relevant security groups of the associated VPC and the endpoint VPC to allow traffic between them
+4) Add the endpoint VPC to HiveConfig.spec.awsPrivateLink.endpointVPCInventory`,
+		Args: cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := opt.Complete(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Validate(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Run(cmd, args); err != nil {
+				return
+			}
+		},
+	}
+
+	regionFlag := "region"
+	subnetIdsFlag := "subnet-ids"
+
+	flags := cmd.Flags()
+	flags.StringVar(&opt.endpointVpcRegion, regionFlag, "", "AWS Region of the endpoint VPC to add")
+	flags.StringSliceVar(&opt.endpointSubnetIds, subnetIdsFlag, []string{}, "IDs of the endpoint subnets to use (as a comma-separated string)")
+
+	_ = cmd.MarkFlagRequired(regionFlag)
+	_ = cmd.MarkFlagRequired(subnetIdsFlag)
+	return cmd
+}
+
+func (o *endpointVPCAddOptions) Complete(cmd *cobra.Command, args []string) error {
+	o.endpointVpcId = args[0]
+
+	// Get controller-runtime dynamic client
+	dynamicClient, err := hiveutils.GetClient()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create controller-runtime client")
+	}
+	o.dynamicClient = dynamicClient
+
+	// Get HiveConfig
+	hiveConfig := &hivev1.HiveConfig{}
+	if err = o.dynamicClient.Get(context.Background(), types.NamespacedName{Name: "hive"}, hiveConfig); err != nil {
+		log.WithError(err).Fatal("Failed to get HiveConfig/hive")
+	}
+	o.associatedVpcs = hiveConfig.Spec.AWSPrivateLink.AssociatedVPCs
```

Comment:
We should validate that `hiveConfig.Spec.AWSPrivateLink` is not nil here.

Reply:
You're right, that's a pointer field. Fixed.